TUTORIAL_UNDERSTANDING_TYPEDEFS.md

```diff
@@ -0,0 +1,248 @@
+
+#### Optional values
+
+A very useful pattern in modern programming language is the use of an optional type (sometimes called `Maybe`) that indicate that a value might be missing. Just like pairs, a lot of programming languages do not feature a "true" optional type (since they require [Algebraic Data Types](https://en.wikipedia.org/wiki/Algebraic_data_type)) but they still have _some_ sort of encoding. Typically, Java features an [`Optional`](https://docs.oracle.com/javase/8/docs/api/java/util/Optional.html) type aimed at avoiding null pointer exceptions, however, since `Optional` is a reference, it can itself be null.
+
+Let's use the `Swift` definition of optional since it's close to a C-like syntax and captures all the necessary information we need to understand the `Optional` type.
+
+```Swift
+enum Optional<T> {
+    case none
+    case some(T)
+}
+```
+
+Which means that an `Optional` type can be one of two values, either `none`, or `some(T)`. The `T` in the second one refers to a type parameter, which mean that a type fits there, we don't know which
+one yet, but any type could work. For example `some(3)` is a valid value with `T = Int`, likewise `some(True)` is also a valid value with `T = Bool`.
+
+**So how many elements does Optional have?**
+
+This one is tricky because there are two answers. Looking at the definition we can reasonably come to the conclusion that there is only two values in the `enum` and therefore only two values for the `Optional` type: `some` and `none`.
+
+Another answer is that it depends on the parameter type. Indeed we already mentionned that `some(3)` and `some(True)` are valid values, if we were to enumerate other values we would have:
+
+- `1` value for `none`
+- since `some(True)` is a value, then `some(False)` also is one
+- since `some(3)` is a value, so is `some(0)`, `some(1)`, and so on are valid values
+
+For the first case, the type parameter could by anything, the only valid value for `none` is `none`.
+
+For the second case, if the type parameter is `Bool` then we have two values for `some`, `some(True)` and `some(False)`. That's the same amount of values that `Bool` has (2), if we add the `none` value that's `2 + 1 = 3` values in total
+
+For the last case, if the type parameter is `Int` then we have as many value for the `some` case as there are `Integer` values. So that's `2^32 = 8'589'934'592`, adding the `none` value that's `8'589'934'592 + 1 = 8'589'934'593` values.
+
+As we see, the pattern is that the number of elements of `Optional` is the number of element of the parameter type plus one. So if we use `T` as the type parameter in our `size` syntax
+we have `size(optional<T>) = size(T) + 1`
+
+#### Either/Result values
+
+Finally we are going to look at a type that reprensent two possible alternative when computing
+a value. In Haskell this is the `Either` type, in Scala this type is called `Try`, in Rust it
```

Comment:
I think Scala's `Try` isn't quite `Either`. Scala actually has (various) `Either` types, see e.g. https://www.scala-lang.org/api/2.9.3/scala/Either.html from the std lib.